File: 93.py
class Solution:
    def spliter(self, s: str, partToSplit: int) -> Generator[str, None, None]:
        if len(s) == 0 or partToSplit < 0:
            return
        if partToSplit == 0:
            sInt = int(s)
            if s[0] == '0':
                if sInt == 0 and len(s) == 1:
                    yield s
            elif 0 < sInt <= 255:
                yield s
            return
        if s[0] == '0':
            for c in self.spliter(s[1:], partToSplit - 1):
                yield '0.' + c
        else:
            for i in range(1, len(s)):
                if int(s[:i]) <= 255:
                    for c in self.spliter(s[i:], partToSplit - 1):
                        yield s[:i] + '.' + c
                else:
                    break

    def restoreIpAddresses(self, s: str) -> List[str]:
        yield from self.spliter(s, 3)

File: sol/93.cpp
struct Solution {
    bool isValidPart(const string& s){
        size_t sSize = s.size();
        if (sSize == 0 || sSize > 3){
            return false;
        }
        if (s[0] == '0'){
            return (sSize == 1);
        }
        int sint = stoi(s);
        return (0 < sint && sint <= 255);
    }
    void recur(const string& s,
               unsigned short ptsToAdd,
               const string& current,
               vector<string>& res){
        if (ptsToAdd == 0){
            if (this->isValidPart(s)){
                res.push_back(current + s);
            }
            return;
        }
        size_t sSize = s.size();
        if (sSize < ptsToAdd + 1){
            return;
        }
        for (size_t l = 1; l <= sSize - ptsToAdd; ++l){
            string ss = s.substr(0, l);
            if (this->isValidPart(ss)){
                this->recur(s.substr(l, string::npos),
                            ptsToAdd - 1,
                            current + ss + ".",
                            res);
            }
        }
    }
    vector<string> restoreIpAddresses(const string& s) {
        vector<string> res;
        this->recur(s, 3, "", res);
        return res;
    }
};
